### PR TITLE
Chore: Typing on "app/livechat/imports/server/rest"

### PR DESCRIPTION
**Pull request by murtaza98**
## Further comments

Derived from https://github.com/RocketChat/Rocket.Chat/pull/24277

### CODE REVIEW

File: packages/rest-typings/src/v1/omnichannel.ts [hunk 1]
```diff
@@ -224,4 +233,325 @@ export type OmnichannelEndpoints = {
...
+	'livechat/room.visitor': {
+		PUT: (params: { rid: string; oldVisitorId: string; newVisitorId: string }) => { room: IOmnichannelRoom };
+	};
+
+	'livechat/transcript': {
+		POST: (params: { token: string; rid: string; email: string; user: any; subject: string }) => { message: string };
+	};
+
+	'livechat/transfer.history/:rid': {
+		GET: (params: { rid: string }) => { history: number[]; count: number; offset: number; total: number };
```
Comment (KevLehman) [inline packages/rest-typings/src/v1/omnichannel.ts:490]: use `PaginatedResult`

File: packages/rest-typings/src/v1/omnichannel.ts [hunk 2]
```diff
@@ -224,4 +233,325 @@ export type OmnichannelEndpoints = {
...
+				searchText?: string;
+				closedChatsOnly?: booleanString;
+				servedChatsOnly?: booleanString;
+			}>,
+		) => PaginatedResult<{ history: IMessage[] }>;
+	};
+
+	'livechat/visitors.autocomplete': {
+		// selector object is of type => { exceptions?: string[]; conditions?: FilterQuery<ILivechatVisitor>; term: string } }
+		GET: (params: PaginatedRequest<{ selector: string }>) => {
```
Comment (KevLehman) [inline packages/rest-typings/src/v1/omnichannel.ts:547]: There's a comment saying that selector is { exceptions: [], conditions: [] }
why you typed it as string? 🙈 
Also, `term` seems to be missing from params

**Review by KevLehman — CHANGES_REQUESTED** [review]

**Comment by KevLehman** [comment]
I know some of the mistakes are mine 🙈 so thanks for fixing them on this PR 🥳

File: apps/meteor/package.json [hunk 3]
```diff
@@ -27,7 +27,7 @@
 		"eslint:fix": "eslint --ext .js,.jsx,.ts,.tsx . --fix",
 		"stylelint": "stylelint \"app/**/*.css\" \"client/**/*.css\" \"app/**/*.less\" \"client/**/*.less\" \"ee/**/*.less\"",
 		"stylelint:fix": "stylelint --fix \"app/**/*.css\" \"client/**/*.css\" \"app/**/*.less\" \"client/**/*.less\" \"ee/**/*.less\"",
-		"typecheck": "cross-env NODE_OPTIONS=\"--max-old-space-size=4092\" tsc --noEmit --skipLibCheck",
+		"typecheck": "cross-env NODE_OPTIONS=\"--max-old-space-size=8184\" tsc --noEmit --skipLibCheck",
```
[the remark this answers is not in the document]
Comment (murtaza98) [reply apps/meteor/package.json:30]: Yeah, I was getting a heap out of memory error for 4GB, hence I doubled the gigs here. Check the build with error [here](https://github.com/RocketChat/Rocket.Chat/runs/6231106841?check_suite_focus=true)

[the remark this answers is not in the document]
Comment (murtaza98) [reply apps/meteor/package.json:30]: On second thought, instead of doubling it, let me check if 5 or 6 or 7 GB works

File: apps/meteor/app/livechat/server/api.js [hunk 4]
```diff
@@ -1,17 +1,17 @@
...
+import '../imports/server/rest/sms';
+import '../imports/server/rest/users';
+import '../imports/server/rest/upload';
+import '../imports/server/rest/inquiries';
+import '../imports/server/rest/rooms';
+import '../imports/server/rest/appearance';
+import '../imports/server/rest/triggers';
+import '../imports/server/rest/integrations';
+import '../imports/server/rest/visitors';
+import '../imports/server/rest/messages.ts';
```
[the remark this answers is not in the document]
Comment (murtaza98) [reply apps/meteor/app/livechat/server/api.js:13]: It's not a new file but rather a renamed file of `../imports/server/rest/visitor.ts`. Although it was named "visitor", it didn't contain any endpoints for visitors & those endpoints were actually placed under `../imports/server/rest/visitor.js` :/ So I've renamed it appropriately

File: packages/rest-typings/src/v1/omnichannel.ts (hunk 2 above)
Comment (murtaza98) [reply packages/rest-typings/src/v1/omnichannel.ts]: :see_no_evil:  Because it is indeed a string of the json stringified object. The term param is also encoded within the same json object. TBH, I didn't like this approach of strigifying the object in query params, it could have been separate individual params....

Comment (murtaza98) [reply packages/rest-typings/src/v1/omnichannel.ts]: It seems this endpoint is only getting used in one place throughout the code base & it is only passing the `term` param - so the other 2 params (i.e. exceptions, conditions) is aren't getting used. So I think it should be safe to deprecate this endpoint in future versions & further optimize the search query.

File: apps/meteor/app/livechat/imports/server/rest/dashboards.ts [hunk 5]
```diff
@@ -106,7 +106,15 @@ API.v1.addRoute(
 			}
 			end = new Date(end);
 
-			const totalizers = getChatsMetrics({ start, end, departmentId });
+			const totalizers = getChatsMetrics({ start, end, departmentId }) as {
```
[the remark this answers is not in the document]
Comment (murtaza98) [reply apps/meteor/app/livechat/imports/server/rest/dashboards.ts]: I tried to use one of the generic types here which re-use the types we've defined in `rest-typings` package so we don't have to re-declare those types in this file. Will that be okay?

File: apps/meteor/app/livechat/imports/server/rest/dashboards.ts [hunk 6]
```diff
@@ -251,7 +266,13 @@ API.v1.addRoute(
 				return API.v1.failure('The "end" query parameter must be a valid date.');
 			}
 			end = new Date(end);
-			const result = findAllChatMetricsByDepartment({ start, end, departmentId });
+
+			const result = findAllChatMetricsByDepartment({ start, end, departmentId }) as {
```
[the remark this answers is not in the document]
Comment (murtaza98) [reply apps/meteor/app/livechat/imports/server/rest/dashboards.ts]: Same goes here:
> I tried to use one of the generic types here which re-use the types we've defined in `rest-typings` package so we don't have to re-declare those types in this file. Will that be okay?

File: packages/rest-typings/src/v1/omnichannel.ts (hunk 2 above)
Comment (murtaza98) [reply packages/rest-typings/src/v1/omnichannel.ts]: https://github.com/RocketChat/Rocket.Chat/pull/25344#issuecomment-1119509496

**Review by KevLehman — DISMISSED** [review]

File: apps/meteor/package.json [hunk 7]
```diff
@@ -27,7 +27,7 @@
 		"eslint:fix": "eslint --ext .js,.jsx,.ts,.tsx . --fix",
 		"stylelint": "stylelint \"app/**/*.css\" \"client/**/*.css\" \"app/**/*.less\" \"client/**/*.less\" \"ee/**/*.less\"",
 		"stylelint:fix": "stylelint --fix \"app/**/*.css\" \"client/**/*.css\" \"app/**/*.less\" \"client/**/*.less\" \"ee/**/*.less\"",
-		"typecheck": "cross-env NODE_OPTIONS=\"--max-old-space-size=4092\" tsc --noEmit --skipLibCheck",
+		"typecheck": "cross-env NODE_OPTIONS=\"--max-old-space-size=5120\" tsc --noEmit --skipLibCheck",
```
[the remark this answers is not in the document]
Comment (murtaza98) [reply apps/meteor/package.json:30]: I think it is because I keep getting heap exceptions for 4GB. Checkout the previous comment here for more info ->https://github.com/RocketChat/Rocket.Chat/pull/25344#discussion_r866737756

File: apps/meteor/app/livechat/imports/server/rest/inquiries.ts [hunk 8]
```diff
@@ -114,7 +101,7 @@ API.v1.addRoute(
 			return API.v1.success(
 				await findInquiries({
 					userId: this.userId,
-					filterDepartment: department,
+					department,
```
[the remark this answers is not in the document]
Comment (murtaza98) [reply apps/meteor/app/livechat/imports/server/rest/inquiries.ts:104]: It should be just `department`. Check the params [here](https://github.com/RocketChat/Rocket.Chat/blob/f57e6b0e5564361f13180c6aa10db99baa360889/apps/meteor/app/livechat/server/api/lib/inquiries.js#L26)

File: packages/rest-typings/src/v1/omnichannel.ts [hunk 9]
```diff
@@ -224,4 +233,333 @@ export type OmnichannelEndpoints = {
...
+	'livechat/business-hour': {
+		GET: (params: { _id: string; type: string }) => {
+			businessHour: ILivechatBusinessHour;
+		};
+	};
+
+	'livechat/analytics/dashboards/conversation-totalizers': {
+		GET: (params: { start: string | Date; end: string | Date; departmentId?: string }) => {
+			totalizers: {
+				title: 'Total_conversations' | 'Open_conversations' | 'Total_messages' | 'On_Hold_conversations' | 'Total_visitors';
```
[the remark this answers is not in the document]
Comment (murtaza98) [reply packages/rest-typings/src/v1/omnichannel.ts]: Is this what you were suggesting? https://github.com/RocketChat/Rocket.Chat/blob/js_to_ts_livechat/imports/server/rest/packages/core-typings/src/IOmnichannelTotalizers.ts

File: apps/meteor/app/livechat/imports/server/rest/dashboards.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (KevLehman) [reply apps/meteor/app/livechat/imports/server/rest/dashboards.ts]: That should 👀

File: apps/meteor/app/livechat/imports/server/rest/dashboards.ts [hunk 10]
```diff
@@ -195,13 +203,17 @@ API.v1.addRoute(
...
+			const startDate = new Date(start);
 
 			if (isNaN(Date.parse(end))) {
 				return API.v1.failure('The "end" query parameter must be a valid date.');
 			}
-			end = new Date(end);
-			const result = findAllChatMetricsByAgent({ start, end, departmentId });
+			const endDate = new Date(end);
+
+			const result = findAllChatMetricsByAgent({ start: startDate, end: endDate, departmentId }) as OperationResult<
```
Comment (KevLehman) [inline apps/meteor/app/livechat/imports/server/rest/dashboards.ts:213]: Don't feel quite comfortable with this casting 😬 why do you need the "OperationResult", if it should be based on the type you're about to return

File: apps/meteor/app/livechat/imports/server/rest/dashboards.ts [hunk 11]
```diff
@@ -245,13 +256,17 @@ API.v1.addRoute(
...
+			const startDate = new Date(start);
 
 			if (isNaN(Date.parse(end))) {
 				return API.v1.failure('The "end" query parameter must be a valid date.');
 			}
-			end = new Date(end);
-			const result = findAllChatMetricsByDepartment({ start, end, departmentId });
+			const endDate = new Date(end);
+
+			const result = findAllChatMetricsByDepartment({ start: startDate, end: endDate, departmentId }) as OperationResult<
```
Comment (KevLehman) [inline apps/meteor/app/livechat/imports/server/rest/dashboards.ts:266]: Same here 😬

File: apps/meteor/app/livechat/imports/server/rest/visitors.ts [hunk 12]
```diff
@@ -1,50 +1,185 @@
...
+			});
+			return API.v1.success(history);
+		},
+	},
+);
 
-			if (!canAccessRoom(room, this.user)) {
-				throw new Error('not-allowed');
+/* deprecated endpoint - "livechat/visitors.autocomplete", use "livechat/visitors.autocompleteByName" instead */
+API.v1.addRoute(
```
Comment (KevLehman) [inline apps/meteor/app/livechat/imports/server/rest/visitors.ts:116]: Can you move this deprecation to its own PR? 🙈

Comment (murtaza98) [reply apps/meteor/app/livechat/imports/server/rest/visitors.ts:113]: Alright Done!! :grimacing:

File: apps/meteor/app/livechat/imports/server/rest/dashboards.ts (hunk 10 above)
Comment (murtaza98) [reply apps/meteor/app/livechat/imports/server/rest/dashboards.ts]: Alright, I declared a new type within this file for the response to avoid using `OperationResult`.

File: apps/meteor/app/livechat/imports/server/rest/inquiries.ts [hunk 13]
```diff
@@ -54,23 +51,15 @@ API.v1.addRoute(
...
-				}
-				return API.v1.success({
-					inquiry: Meteor.runAsUser(this.bodyParams.userId || this.userId, () =>
-						Meteor.call('livechat:takeInquiry', this.bodyParams.inquiryId),
-					),
-				});
-			} catch (e) {
-				return API.v1.failure(e);
+		async post() {
+			if (this.bodyParams.userId && !(await Users.findOneById(this.bodyParams.userId, { fields: { _id: 1 } }))) {
```
Comment (ggazzo) [inline apps/meteor/app/livechat/imports/server/rest/inquiries.ts:55]: requires some mechanism to validate if userId is string

File: apps/meteor/app/livechat/imports/server/rest/integrations.ts [hunk 14]
```diff
@@ -5,12 +5,10 @@ API.v1.addRoute(
...
-				findIntegrationSettings({
-					userId: this.userId,
-				}),
-			);
+		async get() {
+			const settings = await findIntegrationSettings({
+				userId: this.userId,
+			});
 
 			return API.v1.success(settings);
```
Comment (ggazzo) [inline apps/meteor/app/livechat/imports/server/rest/integrations.ts:13]:
```suggestion
		async get() {
			return API.v1.success(await findIntegrationSettings({
				userId: this.userId,
			}));
```

File: packages/core-typings/src/IOmnichannelTotalizers.ts [hunk 15]
```diff
@@ -0,0 +1,83 @@
+export interface IOmnichannelTotalizers {
```
Comment (ggazzo) [inline packages/core-typings/src/IOmnichannelTotalizers.ts:1]: please move create an omnichannel folder and then put this file inside

**Review by ggazzo — CHANGES_REQUESTED** [review]

File: apps/meteor/app/livechat/imports/server/rest/inquiries.ts [hunk 16]
```diff
@@ -54,23 +52,20 @@ API.v1.addRoute(
...
+			});
+
+			const { userId, inquiryId } = this.bodyParams;
+
+			if (userId && !(await Users.findOneById(userId, { fields: { _id: 1 } }))) {
+				return API.v1.failure('The user is invalid');
 			}
+			return API.v1.success({
+				inquiry: Meteor.runAsUser(userId || this.userId, () => Meteor.call('livechat:takeInquiry', inquiryId)),
+			});
```
Comment (ggazzo) [inline apps/meteor/app/livechat/imports/server/rest/inquiries.ts:68]: is that correct? I know its not related with the pull request, but I can use a different userId, without any permission?

File: apps/meteor/app/livechat/imports/server/rest/facebook.ts [hunk 17]
```diff
@@ -19,51 +19,49 @@ import { Livechat } from '../../../server/lib/Livechat';
...
-			};
+			return API.v1.failure('Invalid signature');
 		}
 
 		const sendMessage = {
 			message: {
-				_id: this.bodyParams.mid,
+				rid: mid,
+				token,
+				msg: text,
```
Comment (ggazzo) [inline apps/meteor/app/livechat/imports/server/rest/facebook.ts:54]: is this a bug or something that is being fixed here?

File: apps/meteor/package.json [hunk 18]
```diff
@@ -261,7 +261,7 @@
 		"fibers": "^5.0.1",
 		"file-type": "^16.5.3",
 		"filenamify": "^4.3.0",
-		"filesize": "^3.6.1",
+		"filesize": "^8.0.7",
```
Comment (ggazzo) [inline apps/meteor/package.json:264]: why is that important for this pull request?

Comment (KevLehman) [reply apps/meteor/package.json:264]: That's because filesize doesn't expose any typings for older versions. And the @types package was not working properly.

Good thing is that, even with the update, our usage of the lib is very basic across RC (we just pass a byte amount and get a string rep) and that API is still there even with the 5 majors updates.

File: apps/meteor/app/livechat/imports/server/rest/facebook.ts [hunk 19]
```diff
@@ -74,26 +72,25 @@ API.v1.addRoute('livechat/facebook', {
 			sendMessage.message.token = visitor.token;
 		} else {
 			sendMessage.message.rid = Random.id();
-			sendMessage.message.token = this.bodyParams.token;
+			sendMessage.message.token = token;
 
 			const userId = Livechat.registerGuest({
 				token: sendMessage.message.token,
```
Comment (KevLehman) [inline apps/meteor/app/livechat/imports/server/rest/facebook.ts:78]: Use `token` instead of `sendMessage.message.token`. It's shorter

File: apps/meteor/app/livechat/imports/server/rest/facebook.ts [hunk 20]
```diff
@@ -19,51 +19,49 @@ import { Livechat } from '../../../server/lib/Livechat';
...
-				success: false,
-				error: 'Invalid signature',
-			};
+			return API.v1.failure('Invalid signature');
 		}
 
 		const sendMessage = {
 			message: {
-				_id: this.bodyParams.mid,
+				rid: mid,
```
Comment (KevLehman) [inline apps/meteor/app/livechat/imports/server/rest/facebook.ts:52]: This should be `_id`, not `rid`

File: apps/meteor/app/livechat/imports/server/rest/facebook.ts [hunk 21]
```diff
@@ -74,26 +72,25 @@ API.v1.addRoute('livechat/facebook', {
...
+				username: undefined,
+				id: undefined,
+				connectionData: undefined,
 			});
 
 			visitor = LivechatVisitors.findOneById(userId);
 		}
 
-		sendMessage.message.msg = this.bodyParams.text;
+		sendMessage.message.msg = text;
```
Comment (KevLehman) [inline apps/meteor/app/livechat/imports/server/rest/facebook.ts:91]: You're setting this on L55, you can remove it here 👀

File: apps/meteor/app/livechat/imports/server/rest/inquiries.ts [hunk 22]
```diff
@@ -29,16 +27,16 @@ API.v1.addRoute(
...
 					name: 1,
 					ts: 1,
 					status: 1,
 					department: 1,
 				},
 			});
-			const totalCount = cursor.count();
-			const inquiries = cursor.fetch();
+			const totalCount = await cursor.count();
+			const inquiries = await cursor.toArray();
```
Comment (KevLehman) [inline apps/meteor/app/livechat/imports/server/rest/inquiries.ts:39]: Do you think a `Promise.all` would be beneficial here? 👀

File: apps/meteor/app/livechat/imports/server/rest/inquiries.ts [hunk 23]
```diff
@@ -54,23 +52,20 @@ API.v1.addRoute(
...
-				return API.v1.failure(e);
+		async post() {
+			check(this.bodyParams, {
+				userId: Match.Optional(String),
+				inquiryId: String,
+			});
+
+			const { userId, inquiryId } = this.bodyParams;
+
+			if (userId && !(await Users.findOneById(userId, { fields: { _id: 1 } }))) {
```
Comment (KevLehman) [inline apps/meteor/app/livechat/imports/server/rest/inquiries.ts:63]: `Users` is raw now, i believe it will ignore the `fields` key (use `projection` if possible)

File: apps/meteor/app/livechat/imports/server/rest/inquiries.ts (hunk 16 above)
Comment (KevLehman) [reply apps/meteor/app/livechat/imports/server/rest/inquiries.ts:68]: I think that's because managers can make other users "take" inquiries.

**Review by KevLehman — CHANGES_REQUESTED** [review]

**Comment by KevLehman** [comment]
Hey @murtaza98 , pls check the conflicts 😬
